### DISCUSSION TITLE
Group tag posts by year

### DISCUSSION THREAD

**Discussion by davidakennedy** [11ty/eleventy#2630] [Q&A] [answered]
Hey everyone,

I haven't done much custom stuff with Eleventy, and am stuck on something. I'd like to group tag posts by year on the tag archive pages. I'm currently doing this with [regular posts, thanks to this code](https://github.com/11ty/eleventy/issues/1284#issuecomment-1026679407). My tag archives are currently [set up like the default setup in Eleventy Base Blog](https://github.com/davidakennedy/davidakennedy.com/blob/main/_src/pages/tags.njk).

Can I use the `postsByYear` custom collection to paginate the tag archive? Or should I create another custom collection?

Thanks for the help!

**Comment by pdehaan** [comment, marked as answer]
Maybe I'm a bit uncertain on what you're trying to build.
Are you saying you want a tag archive page with year separators?
Or are you trying to do some more complex pagination where each tag archive has sub pages for each year's posts?

**Reply by davidakennedy** [reply]
Thanks for the reply, @pdehaan!

> Are you saying you want a tag archive page with year separators?

Yes, sorry if I was unclear there. So something like this on a tag archive page:

```
Tagged “Accessibility”
2022
- Post with tag accessibility

2021
- Post with tag accessibility
```

**Reply by pdehaan** [reply]
Ah, much easier than trying to do complex pagination.
I think the logic would be quite similar to the custom collection (although I'd be reluctant to create X new collections just for yearly groupings). I might go with a custom filter route which takes an array of pages (like an existing collection) then just does the yearly grouping and returns an array of year+posts or something.

So assuming this is your current `postsByYear` logic...
```js
config.addCollection("postsByYear", (collection) => {
  const posts = collection.getFilteredByTag('post').reverse();
  const years = posts.map(post => post.date.getFullYear());
  const uniqueYears = [...new Set(years)];

  const postsByYear = uniqueYears.reduce((prev, year) => {
    const filteredPosts = posts.filter(post => post.date.getFullYear() === year);

    return [
      ...prev,
      [year, filteredPosts]
    ]
  }, []);

  return postsByYear;
});
```

It'd be a custom filter which accepts an array of posts and does "things".
Here's some rough, untested pseudo code:

```js
config.addFilter("filteredPostsByYear", (posts = []) => {
  const years = posts.map(post => post.date.getFullYear());
  const uniqueYears = [...new Set(years)];

  const postsByYear = uniqueYears.reduce((prev, year) => {
    const filteredPosts = posts.filter(post => post.date.getFullYear() === year);
    return [
      ...prev,
      [year, filteredPosts]
    ]
  }, []);

  return postsByYear;
});
```

Then maybe something like:
```njk
{%- set postsByYear = collections[tag] | filteredPostsByYear -%}
```

I might even take it a bit further and try and reuse that filter code within the collection code to keep it a bit DRYer (since it's 99% the same), but that might be going too far.

**Reply by pdehaan** [reply]
I'd have to try building it all locally to give a better (and tested) idea.
You might also be able to do it with 1-2 fewer loops, since it looks like you're looping over `posts[]` quite a bit (once just to get unique years, then looping over _n_ unique years, where each iterations loops through the entire `posts` array again and again to do the re-filtering). It might be possible to do that with a single loop over `posts`, but that might be a gross over optimization depending on how many posts your site [potentially] has.

---

**UPDATE:** Not sure if my math is correct, but my theory was your site has 1000 posts written over 5 years.
So you have to loop over all 1000 posts to get a unique Set of [5] years.
Then you loop over 5 years, and each time loop over all 1000 posts to do secondary filtering (so another 5000 iterations).

So 1000 iterations + 5 iterations + 5*1000 iterations = 6005?

**Reply by pdehaan** [reply]
Ah, was a bit tricker than I expected, but here was my rough attempt at creating a year/post map with a single loop (although it also uses `Object.values()` and `Array#sort()`, so maybe it's silly):

```js
/**
 * @typedef {import('@11ty/eleventy/src/UserConfig')} EleventyConfig
 * @typedef {ReturnType<import('@11ty/eleventy/src/defaultConfig')>} EleventyReturnValue
 * @type {(eleventyConfig: EleventyConfig) => EleventyReturnValue}
 */
module.exports = function (eleventyConfig) {
  eleventyConfig.addFilter("groupByYear", function (pages = []) {
    const pagesMap = {};
    for (const page of [...pages]) {
      const pageYear = page.date.getFullYear();
      const yearlyPosts = pagesMap[pageYear] || [pageYear, []];
      yearlyPosts[1].push(page);
      pagesMap[pageYear] = yearlyPosts;
    }

    return Object.values(pagesMap)
      // Sort the year map in descending order.
      .sort(([year1], [year2]) => year2 - year1);
  });

  eleventyConfig.addFilter("toDateString", function (date) {
    return new Date(date).toLocaleDateString();
  });

  eleventyConfig.addFilter("inspect", require("node:util").inspect);

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

### USAGE (LiquidJS)

```liquid
{%- assign groupedPosts = collections.posts | reverse | groupByYear -%}
{% for arr in groupedPosts %}
  {%- assign year = arr[0] -%}
  {%- assign posts = arr[1] -%}
  <section data-year="{{ year }}" data-count="{{ posts.length }}">
    <h1>{{ year }}</h1>
    <ul>
    {% for post in posts %}
      <li>{{ post.data.title }} on {{ post.date | toDateString }}</li>
    {% endfor %}
    </ul>
  </section>
{% endfor %}
```

### USAGE (Nunjucks)

```njk
{% for year, posts in collections.posts | reverse | groupByYear %}
  <section data-year="{{ year }}" data-count="{{ posts.length }}">
    <h1>{{ year }}</h1>
    <ul>
    {% for post in posts -%}
      <li>{{ post.data.title }} on {{ post.date | toDateString }}</li>
    {% endfor -%}
    </ul>
  </section>
{% endfor %}
```

### OUTPUT

```html
  <section data-year="2023" data-count="1">
    <h1>2023</h1>
    <ul>
      <li>ThReE on 2023-03-12</li>
    </ul>
  </section>

  <section data-year="2022" data-count="2">
    <h1>2022</h1>
    <ul>
      <li>tWo on 2022-02-21</li>
      <li>oNe on 2022-01-30</li>
    </ul>
  </section>

  <section data-year="2016" data-count="1">
    <h1>2016</h1>
    <ul>
      <li>SeVeN on 2016-06-26</li>
    </ul>
  </section>

  <section data-year="2015" data-count="2">
    <h1>2015</h1>
    <ul>
      <li>SiX on 2015-06-15</li>
      <li>FiVe on 2015-05-24</li>
    </ul>
  </section>

  <section data-year="2014" data-count="2">
    <h1>2014</h1>
    <ul>
      <li>EiGhT on 2014-08-17</li>
      <li>FoUr on 2014-04-13</li>
    </ul>
  </section>
```

**Reply by davidakennedy** [reply]
Huge thanks, @pdehaan! I was able to use this code as intended for both my [blog archive](https://davidakennedy.com/blog/) and tag archive ([example tag](https://davidakennedy.com/tag/accessibility/)).

I don't think I would have gotten to a solution like this on my own. Still learning the ins and outs of both JavaScript and how Eleventy manages data. It's super elegant too since I can use this with any collection. Going to study this more so I can learn!

**Reply by pdehaan** [reply]
Bah, you already had all the code, I just shuffled it slightly. #teamwork

**Reply by pdehaan** [reply]
OH, i forgot to update on reusing filters...
```js
// .eleventy.js snippet
  eleventyConfig.addFilter("groupByYear", function (pages = []) {
    const pagesMap = {};
    for (const page of [...pages]) {
      const pageYear = page.date.getFullYear();
      const yearlyPosts = pagesMap[pageYear] || [pageYear, []];
      yearlyPosts[1].push(page);
      pagesMap[pageYear] = yearlyPosts;
    }

    return Object.values(pagesMap)
      // Sort the year map in descending order.
      .sort(([year1], [year2]) => year2 - year1);
  });

  eleventyConfig.addCollection("thingsByYear", function (collectionApi) {
    const groupByYear = eleventyConfig.getFilter("groupByYear");
    const pages = collectionApi.getFilteredByTag("thing");
    return groupByYear(pages);
  });
```

So above your custom collection can reuse your `groupByYear` filter by using the `eleventyConfig.getFilter()` API. Makes the code a bit more reusable if you need to share the same code between filters and collections, or multiple custom collections, etc.

Or you can go a bit further and create a helper function for your custom collections which takes a tag name and returns the collection creation function:

```js
  function addGroupedCollection(tag = "") {
    const groupByYear = eleventyConfig.getFilter("groupByYear");
    return (collectionApi) => {
      const pages = [...collectionApi.getFilteredByTag(tag)];
      return groupByYear(pages);
    };
  }
  eleventyConfig.addCollection("a11yByYear", addGroupedCollection("a11y"));
```

Possibly not helpful or useful in this exact case, but can be nice to know that it's an option if you need it in <em>the future<sup>(tm)</sup></em>.

**Reply by davidakennedy** [reply]
Super helpful, thanks for the explanations!
